File: example/templates/addons/desktop_widget/weather.py
#!/usr/bin/env python3
"""
MeowterialYou Weather Helper
Outputs plain weather data (one value per line) for Conky to format
"""

import math
import os
import gi
gi.require_version('GWeather', '4.0')
gi.require_version('Gio', '2.0')
from gi.repository import GWeather, Gio, GLib

CACHE_FILE = os.path.expanduser('~/.cache/meowterialyou_weather')
CACHE_DURATION = 900
APP_ID = 'io.github.meowterialyou.widget'


def get_cached():
    import time
    try:
        if os.path.exists(CACHE_FILE):
            if time.time() - os.path.getmtime(CACHE_FILE) < CACHE_DURATION:
                with open(CACHE_FILE, 'r') as f:
                    return f.read().strip()
    except Exception:
        pass
    return None


def save_cache(data):
    try:
        os.makedirs(os.path.dirname(CACHE_FILE), exist_ok=True)
        with open(CACHE_FILE, 'w') as f:
            f.write(data)
    except Exception:
        pass


def get_gnome_weather_location():
    try:
        settings = Gio.Settings.new('org.gnome.Weather')
        locations = settings.get_value('locations')
        
        if locations.n_children() > 0:
            child = locations.get_child_value(0)
            inner = child.get_variant()
            loc_data = inner.get_child_value(1).get_variant()
            city = loc_data.get_child_value(0).get_string()
            coords_array = loc_data.get_child_value(3)
            
            if coords_array.n_children() > 0:
                coord = coords_array.get_child_value(0)
                lat = math.degrees(coord.get_child_value(0).get_double())
                lon = math.degrees(coord.get_child_value(1).get_double())
                return city, lat, lon
    except Exception:
        pass
    return None, None, None


def get_weather_icon_char(icon_name):
    """Map GNOME weather icon to Nerd Font character."""
    if not icon_name:
        return ''  # sunny
    
    icon_name = icon_name.lower()
    if 'clear' in icon_name and 'night' in icon_name:
        return ''  # night clear
    elif 'clear' in icon_name or 'sunny' in icon_name:
        return ''  # sunny
    elif 'few-clouds' in icon_name or 'partly' in icon_name:
        return ''  # partly cloudy
    elif 'overcast' in icon_name or 'cloud' in icon_name:
        return ''  # cloudy
    elif 'fog' in icon_name or 'mist' in icon_name:
        return ''  # fog
    elif 'shower' in icon_name:
        return ''  # showers
    elif 'rain' in icon_name:
        return ''  # rain
    elif 'snow' in icon_name:
        return ''  # snow
    elif 'storm' in icon_name or 'thunder' in icon_name:
        return ''  # storm
    else:
        return ''  # default sunny


def get_weather():
    cached = get_cached()
    if cached:
        print(cached)
        return

    city, lat, lon = get_gnome_weather_location()
    
    if not city or lat is None:
        # Output placeholder values
        print("--")  # temp
        print("Unknown")  # condition
        print("Unknown")  # city
        print("--")  # wind
        print("--")  # humidity
        print("")  # icon
        return

    location = GWeather.Location.new_detached(city, None, lat, lon)
    info = GWeather.Info.new(location)
    info.set_application_id(APP_ID)
    info.set_contact_info('meowterialyou@widget')
    info.set_enabled_providers(GWeather.Provider.MET_NO | GWeather.Provider.OWM)

    loop = GLib.MainLoop()
    result = [None]

    def on_updated(info, *args):
        try:
            temp_ok, temp = info.get_value_temp(GWeather.TemperatureUnit.CENTIGRADE)
            sky = info.get_sky() or ""
            icon_name = info.get_icon_name() or ""
            humidity = info.get_humidity() or ""
            wind = info.get_wind() or ""
            
            # Format values
            temp_str = f"{temp:.0f}°" if temp_ok else "--°"
            
            # Get condition from icon name if sky is empty
            if not sky or sky == "-":
                if 'clear' in icon_name.lower():
                    sky = "Clear"
                elif 'cloud' in icon_name.lower():
                    sky = "Cloudy"
                elif 'rain' in icon_name.lower():
                    sky = "Rainy"
                elif 'snow' in icon_name.lower():
                    sky = "Snowy"
                else:
                    sky = "Clear"
            
            city_str = city[:15] if len(city) > 15 else city
            
            # Parse humidity
            if humidity and humidity != "-":
                hum_str = humidity.replace('%', '').strip()
            else:
                hum_str = "--"
            
            # Parse wind
            if wind and wind != "-":
                if "/" in wind:
                    wind_str = wind.split("/")[1].strip().replace(" ", "")
                else:
                    wind_str = wind.strip()
            else:
                wind_str = "--"
            
            # Get icon
            icon_char = get_weather_icon_char(icon_name)
            
            # Output: one value per line
            # Line 1: temperature
            # Line 2: condition
            # Line 3: city
            # Line 4: wind
            # Line 5: humidity
            # Line 6: weather icon character
            output = f"{temp_str}\n{sky}\n{city_str}\n{wind_str}\n{hum_str}\n{icon_char}"
            result[0] = output
            
        except Exception as e:
            result[0] = f"--\nError\n{city}\n--\n--\n"
        finally:
            loop.quit()

    info.connect('updated', on_updated)
    info.update()
    GLib.timeout_add_seconds(12, loop.quit)
    loop.run()

    if result[0]:
        save_cache(result[0])
        print(result[0])
    else:
        print(f"--\nUnknown\n{city}\n--\n--\n")


if __name__ == '__main__':
    try:
        get_weather()
    except Exception:
        print("--\nError\nUnknown\n--\n--\n")

File: example/templates/addons/desktop_widgets/weatherclock/src/services/weather.ts

import GWeather from 'gi://GWeather?version=4.0';
import GLib from 'gi://GLib?version=2.0';
import Gio from 'gi://Gio?version=2.0';
// @ts-ignore
import { log } from '../utils.js';
import { Config } from '../config.js';

let _info: GWeather.Info;
let _location: GWeather.Location;

// Helper to get location from GNOME Shell settings directly
function getGnomeWeatherLocation(): [string, string | null, number, number] | null {
  try {
    const settings = new Gio.Settings({ schema_id: 'org.gnome.Weather' });
    const locations = settings.get_value('locations');

    if (locations.n_children() > 0) {
      const child = locations.get_child_value(0);
      const inner = child.get_variant();
      const locData = inner.get_child_value(1).get_variant();

      const city = locData.get_child_value(0).get_string()[0];
      let code: string | null = null;
      try {
        code = locData.get_child_value(1).get_string()[0];
      } catch (e) {}

      const coordsArray = locData.get_child_value(3);
      if (coordsArray.n_children() > 0) {
        const coord = coordsArray.get_child_value(0);
        const lat = coord.get_child_value(0).get_double();
        const lon = coord.get_child_value(1).get_double();
        return [city, code, lat, lon];
      }
    }
  } catch (e) {
    // log(`Error reading GNOME Weather location: ${e}`);
  }
  return null;
}

export function getWeatherIconChar(iconName: string): string {
  if (!iconName) return '󰖙';
  const lower = iconName.toLowerCase();
  if (lower.includes('clear') && lower.includes('night')) return '';
  if (lower.includes('clear') || lower.includes('sunny')) return '';
  if (lower.includes('few-clouds') || lower.includes('partly')) return '󰖕';
  if (lower.includes('overcast') || lower.includes('cloud')) return '󰖐';
  if (lower.includes('fog') || lower.includes('mist')) return '󰖑';
  if (lower.includes('shower')) return '󰖖';
  if (lower.includes('rain')) return '󰖗';
  if (lower.includes('snow')) return '󰖘';
  if (lower.includes('storm') || lower.includes('thunder')) return '󰖓';
  return '';
}

export const initWeather = (config: Config, onUpdate?: () => void) => {
   const sysLoc = getGnomeWeatherLocation();
   if (sysLoc) {
       const [city, code, lat, lon] = sysLoc;
       log(`Found system weather location: ${city}`);
       _location = GWeather.Location.new_detached(city, code, lat, lon);
   } else {
       log(`System weather location not found, falling back to Pokhara`);
       _location = GWeather.Location.new_detached('Pokhara', 'VNPK', 28.2096 * (Math.PI / 180.0), 83.9856 * (Math.PI / 180.0));
   }
   
   _info = new GWeather.Info({
       location: _location,
       application_id: 'meowterialyou.widget',
       contact_info: 'https://github.com/meowterialyou',
   });
   _info.set_enabled_providers(GWeather.Provider.MET_NO);
   
   _info.connect('updated', () => {
       if (onUpdate) onUpdate();
   });
   _info.update();
};

export const forceWeatherUpdate = () => {
    if (_info) _info.update();
};

export const getWeather = (config: Config) => {
    if (!_info) return { temp: '--', iconChar: '', desc: '...', city: '...', humidity: '', wind: '' };

    const unit = config.weather.unit === 'F' ? GWeather.TemperatureUnit.FAHRENHEIT : GWeather.TemperatureUnit.CENTIGRADE;
    const [ok, temp] = _info.get_value_temp(unit);
    const iconName = _info.get_icon_name();
    const summary = _info.get_weather_summary();

    // @ts-ignore
    const humidity = _info.get_humidity ? _info.get_humidity() : '';
    
    // Wind
    let windStr = '';
    try {
        const speedUnit = config.weather.wind_unit === 'mi' ? GWeather.SpeedUnit.MPH : GWeather.SpeedUnit.KPH;
        const [windOk, windSpeed, windDirEnum] = _info.get_value_wind(speedUnit);
        if (windOk) {
            const unitLabel = config.weather.wind_unit === 'mi' ? 'mph' : 'km/h';
            const rawDir = GWeather.wind_direction_to_string(windDirEnum);
            windStr = `${Math.round(windSpeed)} ${unitLabel} ${rawDir || ''}`;
        }
    } catch(e) {}

    const unitSymbol = config.weather.unit === 'F' ? '°F' : '°';
    const tempStr = ok ? `${Math.round(temp)}${unitSymbol}` : `--${unitSymbol}`;
    const iconChar = getWeatherIconChar(iconName || '');
    
    let desc = summary || '';
    if (!desc || desc.toLowerCase().includes('unknown') || desc === '??') {
         // Fallback based on icon
         if (iconName) {
            const lowerIcon = iconName.toLowerCase();
            if (lowerIcon.includes('clear') || lowerIcon.includes('sunny')) desc = 'Clear';
            else if (lowerIcon.includes('cloud')) desc = 'Cloudy';
            else if (lowerIcon.includes('rain')) desc = 'Rain';
            else desc = '...';
         } else {
             desc = '...';
         }
    }
    
    // Graceful error handling
    if (desc.toLowerCase().includes('failed') || desc.toLowerCase().includes('error')) {
        desc = 'Offline';
    }

    // Clean up desc
    if (_location) {
        const locName = _location.get_name();
        if (locName && desc.includes(locName)) {
            desc = desc.replace(locName, '').trim();
        }
    }
    if (desc.length > 20) desc = desc.substring(0, 20) + '...';

    const city = _location ? _location.get_name() : 'Unknown';

    return { temp: tempStr, iconChar, desc, city, humidity, wind: windStr };
};
